### ISSUE TITLE
LUKS encrypted disks mounted with pam_mount fail to unmount on shutdown

### ISSUE THREAD

**Issue by BowDown097** [davmac314/dinit#191]
Shutdown process freezes on "Unmounting disks..." if any LUKS encrypted disks have been mounted with pam_mount. May extend to other types of encryption as well, but I can't really test anything outside of LUKS specifically. Nothing related to the issue seems to be present in any log files - is there any better way to debug this?

Here's my config to unmount the drives if that helps:
```xml
<volume user="bow" fstype="crypt" path="/dev/disk/by-uuid/c306bf2b-e180-4699-8b3b-1a5d3e755978" mountpoint="/run/media/bow/USB" options="crypto_name=USB,allow_discard,uid=bow,gid=bow" />
<volume user="bow" fstype="crypt" path="/dev/disk/by-uuid/54f1d498-302e-4aeb-948b-5fcebe8fda81" mountpoint="/media/sdb1" options="crypto_name=Data,allow_discard" />
```

**Comment by davmac314** [comment]
"Unmounting disks" is performed by executing "umount -a -r" (umount is from the util-linux package). Output is captured and should be displayed, so if it's not displaying anything then my guess is that `umount` itself is hanging.

I suggest setting up a shutdown hook (see man page for shutdown, source is [here](https://github.com/davmac314/dinit/blob/master/doc/manpages/shutdown.8.m4)) which starts up a terminal on a tty (see the tty services for example command). This will allow you to try `umount -a` yourself (to confirm this is the problem) and try anything else that you want to.

So in short: my guess is that pam_mount is doing something that umount isn't able to handle. It might well be a bug in umount.

**Comment by davmac314** [comment]
I'm not sure it is dinit-specific, see https://github.com/systemd/systemd/issues/2351 and https://github.com/systemd/systemd/issues/2351

Since the `umount` command hangs, that really implies to me that there's a bug in `umount`. Dinit isn't involved in that operation.

**Comment by davmac314** [comment]
I will close this now, I suggest opening a bug against util-linux if it remains a problem.

**Comment by davmac314** [comment]
@dkwo as I understand it, if root is mounted on an encrypted volume, `cryptsetup close` isn't actually necessary. All that `close` does is remove device mapping (and purge the key from memory). The usual process of remounting root as read-only, sync'ing disk and then shutting down the system should suffice. Omitting `cryptsetup close` doesn't have any effect that lasts beyond shutdown.

> This may have implications e.g. for nvme ssd drive's health.

Why do you say that? I'm not following.
